designs/aws/aws-launch-templates-v2.md

```diff
@@ -0,0 +1,163 @@
+* It most directly addresses what every customer is asking - let them configure EC2 properties from within the provisioner.
+
+*Cons*
+* APIs should really never accept weakly typed input. It's going to be very easy for a customer to get something wrong what went wrong break the deserialization. When that happens, it might be equally difficult to debug YAML.
+* If customers have to fill out the whole launch templates, it's going to make the provisionerSpec verbose. If they only fill it out partially, then you have the same cons as approach 2.
+
+
+**Approach 4 - Let all infrastructure configuration be expressed as a pod requirement**
+
+The essence of this approach, is to represent each EC2 instance level feature as a well known label that the Karpenter provisioner recognizes. This is what we currently do with instanceTypes with the `node.kubernetes.io/instance-type` label. Using this approach, the provisioner can read off the podSpec to determine which ImageId it needs to use, which additional volumes need to be statically provisioned and so on.
+
+*Pros*
+* Karpenter supports a kube-native way to define node level configuration that is agnostic of the cloud provider.
+* Since each application is unique and may have unique infrastructure requirements, users will be otherwise forced to define multiple provisioners where each provisioner uses a distinct launch template. By moving the configuration to the pod level, we can continue to only necessitate a single / fewer provisioners.
+
+
+*Cons*
+* Labels aren't expressive enough for free-form fields like UserData or even complex fields like say volumes that have multiple nested attributes.
+* Karpenter uses layered constraints - is it possible to define a bound for UserData?
+  * Along the same lines, should the application developer always have complete control over their infrastructure? Some configuration like ImageIds might need to be enforced by the cluster administrator, so it would be beneficial to never expose ImageIds as a knob to the app developer at all.
+* By moving instance configuration from the provider label to the provisioner label, each separate cloud provider implementation will need to support that label which introduces unneeded complexity. The feature set across all cloud providers is unlikely to ever be homogenous.
+
+
+### Recommendation
+
+I recommend we continue with Approach 1, and give everyone the ability to customize a few select, commonly used fields in the provisionerSpec and ask them to configure all others via a complete LaunchTemplate. I think this keeps our API easy to use. The fields I expect us to add to the Provider in the short term are _BlockDeviceMappings, ImageId / AMIFamily and UserData._ The use cases for other fields either don’t apply to Karpenter, or they’re quite niche where it probably makes sense for the customer to manage all other aspects of the launch template in any case.
```

Comment:
Personally, I am more in favor of Approach 3 as it is more flexible (allows for more use-cases) without having to change karpenter's code base again and again. I am looking at this from an operator's (not developer's) point-of-view, in which it is often frustrating and annoying that developers do not understand that operator's (customer's) need for flexible tools (and not just opinionated and limited tools that don't get the job done).

Reply:
I try and call this out early in the doc, but the core reason I'm recommending the first approach is because we often see a need to only override *one* or *two* of the AWS Provider's opinionated defaults, so defining an entire launch template is too heavy handed. 

But if you've got more advanced needs that frequently change, maintaining an entire launch template shouldn't be too cumbersome and likely similar to embedding all of that information within the provisioner.